crates/codegen/parser/runtime/src/query_model.rs

```diff
@@ -0,0 +1,144 @@
+use std::rc::Rc;
+
+use super::kinds::{FieldName, RuleKind, TokenKind};
+
+#[derive(Clone)]
+pub enum Query {
+    Binding(Rc<BindingMatch>),
+    Node(Rc<NodeMatch>),
+    Optional(Rc<OptionalMatch>),
+    Alternatives(Rc<AlternativesMatch>),
+    Sequence(Rc<SequenceMatch>),
+    OneOrMore(Rc<OneOrMoreMatch>),
+    Ellipsis,
+}
+
+impl Query {
+    pub fn parse(text: &str) -> Result<Self, String> {
+        super::query_parser::parse_query(text)
+    }
+
+    pub fn unparse(&self) -> String {
```

Comment:
TOL, since `unparse()` doesn't store/reconstruct the input query, what do you think about using something like `impl Display`>`to_string()` instead?
Would also make it a bit more idiomatic/effecient, as it uses a single writer instead of creating/dropping temporary strings.

Reply:
Good reasoning, done